include/odp/api/spec/traffic_mngr.h

```diff
@@ -605,6 +605,42 @@ odp_tm_t odp_tm_find(const char            *name,
  */
 int odp_tm_capability(odp_tm_t odp_tm, odp_tm_capabilities_t *capabilities);
 
+/**
+ * Start a TM system
+ *
+ * odp_tm_start() needs to be used to start an already created or found TM
+ * system. By default, all the TM systems are in stopped state.
+ *
+ * @param tm  TM system to be started
+ *
+ * @retval 0  on success
+ * @retval <0 on failure
+ */
+int odp_tm_start(odp_tm_t tm);
+
+/**
+ * Stop a TM system
+ *
+ * odp_tm_stop() can to be used to stop a TM system that is already started.
+ * Stopping a TM system will cause it to stop / pause processing packets and
+ * become idle.
```

Comment:
The API text here may contradict with the later text or at least be confusing. Here it is said that the TM systems stops processing packets and becomes idle, but the description below says that the TM may continue transmitting packets (but without QoS treatment).

Then it looks like the linux-gen implementation, which supports dynamic updates and implements odp_tm_stop() as a no-op, may be in violation of this API spec. Linux gen TM in a stopped state will continue processing packets with QoS treatment, although that is not mentioned as a possibility in the text. Also, the TM does not become idle if it is fed enough packets (e.g. by another thread), so the promise about becoming idle is not kept.

I am not sure to what extent this problem is about the API text being a bit vague and not fully communicating the real intent and to what extent the intended API needs rethinking.

One possibility would be to define odp_tm_stop() like this:

```
 * Stop a TM system
 *
 * odp_tm_stop() can to be used to stop a TM system that is already started.
 * When a TM system is in a stopped state, its configuration can be freely
 * modified but it is undefined whether the TM system is processing packets
 * normally, processing packets without TM treatment, dropping packets silently
 * or returning an error to possible enqueue operation.
 *
 * Stopping a TM system is needed for changing TM configuration for which
 * dynamic updates are not supported. It is undefined whether TM configuration
 * changed done in the stopped stae take effect immediately or when the
 * TM system is started.
 *
 * A TM system must be stopped before it is destroyed.
```

And the description of odp_tm_stop() could be amended with this: "TM configuration changes for which dynamic updating is not supported may not take effect until the TM is started."

I am not sure if the above API is any good but I think it probably matches the intended behaviour as I have understood it. It also makes no claims regarding what odp_tm_is_idle() returns for a stopped TM.

The above API looks a bit ugly to me, partly because it leaves so much unspecified and partly because it calls the functions odp_tm_start() and odp_tm_stop() even though there is no promise that a TM in a stopped state is really stopped in any way. So odp_tm_stop() is actually more like odp_tm_prepare_for_config_change() and odp_tm_start() is really odp_tm_config_changes_done().

Reply:
I also feel the above text you proposed has lot of undefined behavior.

With existing text, I think since there is an option of processing traffic without TM treatment when in stopped state provided to platforms, I can remove "become idle" part of the text from API to resolve the issue.

And for linux-gen, I can set and clear a flag in odp_tm_start() and odp_tm_stop() and change the odp_tm_enq() to fail if TM is in stopped state if it is fine.

Reply:
Or, remove the API text that says "New packets will be transmitted without TM treatment" as after fixing linux-generic, packets will not be transmitted. Our platform also doesn't support transmission in stopped state

Reply:
I had a chat with @psavol about this and now we both think that the API should say that when a TM is in the stopped state, an ODP application must not enqueue new packets to it either directly through odp_tm_enq() and odp_tm_enq_with_cnt() or indirectly through odp_ipsec_out_inline(). Thus the application must call odp_tm_start() before calling those functions and must stop calling those functions (for the relevant TM) before calling odp_tm_stop(),

The fate of packets that are in flight, either inside TM or inside IPsec, could be left unspecified: they might get sent out with unspecified TM treatment or they might get dropped.

How does that sound?

Reply:
I am not sure what you mean with linux-generic issue. If you are referring to the fact that the linux-generic code in this PR does not drop packets at tm_enqueue() when the TM is in the stopped state, there is no problem with that with the revised API since an application would not anymore be allowed to call tm_enqueue() when the TM is in the stopped state.

Reply:
Yes, I meant the same. Apologies for not describing it better.

Ack. I'll leave the linux-generic code for odp_tm_start() and odp_tm_stop() as it is.

Reply:
The description of odp_tm_stop() still promises too much as it talks about processing packets and the TM becoming idle although the TM may still continue processing in-flight packets. I would propose changing the first paragraph to something like this:

"odp_tm_stop() can to used to stop a TM system that is already started for the purpose of reconfiguration that cannot be done dynamically."

Reply:
Ack.